drafts/use-cases/2.api-documentation.md

```diff
@@ -0,0 +1,111 @@
+So it can pre-generate some views and other UI elements to speed things up.
+
+### Usage
+```javascript
+var client = new HydraClient();
+var apiDocumentation = client.getApiDocumentation("/api");
+for (var $class of apiDocumentation.supportedClasses) {
+    // generate a partial view for the _$class_
+}
+```
+
+### Details
+The documentation obtain may be a large payload,
+thus client is expected to store it temporarely at least for the current session.
+
+```
+GET /api?documentation
+
+HTTP 200 OK
+```
+```javascript
+{
+    "@id": "/api?documentation",
+    "supportedClasses": [
+        {
+            "@id": "schema:Event",
+            "suportedProperties": [
+                {
+                    "property": {
+                        "@id": "eventName"
+                    }
+                },
+                {
+                    "property": {
+                        "@id": "eventDescription"
+                    }
+                },
+                {
+                    "property": {
+                        "@id": "startDate"
+                    }
+                },
+                {
+                    "property": {
+                        "@id": "endDate"
+                    }
+                }
+            ]
+        }
+    ]
+}
+```
+
+In order to i.e. generate views for each supported class, a complete type definition should be provided.
+Currently, Hydra Core Vocabulary allows that using RDFS/OWL vocabularies.
+
+The documentation should also contain some human readable details like:
+- operation expected type and it's role
+- operation returned type and it's role
+- operation name
+- operation description
+- parameters and their descriptions
+
+### Considerations
+
+#### Entry-Point/API Documentation fetching precedence
+In the previous scenario, application fetched an entry-point data as it's first action.
+We also could consider an opposite - taking an API documentation if available.
+This would require some standardized mechanism of accesing that documentation.
+
+Few possibilities here are available:
+- sending an _OPTIONS_ HTTP request to the base URL,
+  but HTTP specification does not allow to cache these responses
+- sending an _GET_ HTTP request to the base URL with some HTTP header;
+  header _Expect_ would be a candidate, but it's not allowed in AJAX communication;
+  another option would be a _Prefer_
+
+Once the API documentation is fetched and processed,
+client would then display some base view with further options,
+i.e. a main menu with all supported classes.
+
+#### Data structure description
+As already mentioned, Hydra Core Vocabulary allows to embed class description using RDFS/OWL constructs.
+Unfortunately, using those may prove difficult in parsing.
+In the scenario above, we'd need to use _rdfs:range_ predicate to denote the types a given property accepts.
+For simple scenarios it may be just _xsd:string_, but what if server will come out with a union of classes?
```

Comment:
I'm not sure I agree. Sure, using OWL constructs could be a problem but the client won't really need to do inferencing in this particular case. For me description like below would be completely fine

```
{
  "suportedProperties": [
    {
      "@id": "ex:title",
      "rdfs:range": "xsd:string"
    },
    {
      "@id": "ex:spouse",
      "rdfs:range": {
        "owl:unionOf": [ "ex:Class1", "ex:Class2" ]
      }
    }
  ]
}
```

In that the `ex:something` accepts either `ex:Class1` or `ex:Class2`. Note however that I used meaningless names because I couldn't come up with a good example.

Nonetheless, I don't think it's too much to expect the client to understand these semantics.